[#37] Update design of confirmation letter

cleac opened the pull request:
Resolves #37 

Implement the updated design of confirmation letter.

@webknjaz  @bluebirrrrd  @anxolerd: give your suggestions of text improvements

cleac committed:
Implementat of new design of confirmation

cleac committed:
Fix underline in `confirm` link

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+        <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
+    </head>
+    <body style="marin: 8px; padding: 0px; max-width: 768px">
```
`marGin`

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <center><img style="width: 200px" width="200px" src="/img/gdg_ukraine_logo.png"></center>
+        <p style="margin-left: 20px; margin-bottom: 20px; font-family: Lato, Arial, sans-serif">Hey, pal!</p>
+        <p style="margin-left: 20px; font-family: Lato, Arial, sans-serif">
+        We are writing to you, because you have registered to "{{ event.title }}". We just want to clarify, whether you are going to attend the event, or you have another plans. You can get more info about an event right <a href="{{}}">here</a>.
```
`{{}}`

webknjaz replied:
+80 chars lines is more readable

webknjaz replied:
not sure about comma before "whether"

webknjaz replied:
+ https://www.usingenglish.com/forum/threads/52133-writing-you-vs-writing-to-you

anxolerd replied:
As for this paragraph, user see no value to himself (and no reason in consequence) to click the button. What about smth like: 

```
We are glad you are interested in {{ event.title }}. 
There is one small step left to complete your registration...
```

We have to give some value to user so he/she could see the reason to click that button, not just ignore this email.

cleac replied:
I don't like how it spells. It's a bit... business-formal or somehow like that.

Also that staff about *one little step*, tells that user has to do more staff for us (while noone likes to do a lot of staff for someone - event just click on a button). And, also, it generates more text he has to read. Do you read all those long emails, the only purpose of which is to confirm that you are *actually* going to go somewhere?

I think, it's better to make it like that:

`begin-part`

Thanks for registering to the event "{{ event.title }}".
We are really looking forward to see you there, but we must clarify that you will come. If you will, press ["I will come"]().

If not, no action is needed, just ignore this letter.

`end-part`

This is kinda draft and has some issues.

webknjaz replied:
http://www.english-for-students.com/What-is-the-difference-between-Staff-and-Stuff.html

webknjaz replied:
knew -> know

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <p style="margin-left: 20px; font-family: Lato, Arial, sans-serif">
+        We are writing to you, because you have registered to "{{ event.title }}". We just want to clarify, whether you are going to attend the event, or you have another plans. You can get more info about an event right <a href="{{}}">here</a>.
+        </p>
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If you are going to attend, press the button below.</p>
```
Again, I don't want to scroll horizontally

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        We are writing to you, because you have registered to "{{ event.title }}". We just want to clarify, whether you are going to attend the event, or you have another plans. You can get more info about an event right <a href="{{}}">here</a>.
+        </p>
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If you are going to attend, press the button below.</p>
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If no, just ignore this email.</p>
```
`noT`

anxolerd replied:
`Otherwise`?

cleac replied:
+1 to `Otherwise`. Seems better for me

webknjaz replied:
both are valid

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If no, just ignore this email.</p>
+        <a href="{{ confirm.url | e }}" style="text-decoration: none">
+            <div style="margin: 10px auto; background-color: blue; width: 160px; height: 35px; color: white; text-align: center; vertical-align: baseline; border-radius: 3px; border-color: blue" width="160px" height="35px" bgcolor="blue">
+                <div style="padding: 6px; font-family: Lato, Arial, sans-serif">I am going to attend</div>
```
attend -> come

anxolerd replied:
Make the button text shorter. Smth like `I'll be there` or `I will come`...

webknjaz commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+    </head>
+    <body style="marin: 8px; padding: 0px; max-width: 768px">
+        <center><img style="width: 200px" width="200px" src="/img/gdg_ukraine_logo.png"></center>
+        <p style="margin-left: 20px; margin-bottom: 20px; font-family: Lato, Arial, sans-serif">Hey, pal!</p>
```
pal -> `rand('dude', 'man', 'pal', etc)`

anxolerd replied:
I am thinking about using participant's name here: 
`Hi {{ u.name or 'friend' }}`

As for `pal`, `dude`, `man` and so on, I believe we should replace it with something less gender-specific. You know, there can be female participants as well... =)

cleac replied:
@anxolerd `pal` is not gender-specific, as I think.... But that's good idea about `{{ u.name or 'placeholder' }}`

webknjaz reviewed (changes requested):
The text may still need some fixes

anxolerd commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        </a>
+        <p style="margin-left: 20px; font-family: Lato, Arial, sans-serif">
+        Sincerely yours,<br>
+        friends from GDG Ukraine
```
We have a group related to GDG event. Probably we should be more specific and use it.

cleac replied:
Truth. Will fix it

anxolerd commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If you are going to attend, press the button below.</p>
+        <p style="margin: 0px 20px; font-family: Lato, Arial, sans-serif">If no, just ignore this email.</p>
+        <a href="{{ confirm.url | e }}" style="text-decoration: none">
+            <div style="margin: 10px auto; background-color: blue; width: 160px; height: 35px; color: white; text-align: center; vertical-align: baseline; border-radius: 3px; border-color: blue" width="160px" height="35px" bgcolor="blue">
```
What about not hardcoding `height`?
![image](https://cloud.githubusercontent.com/assets/6533582/18611082/c5e6528c-7d37-11e6-88de-a65110b3425b.png)

cleac replied:
Thx for checking. I think I'll fix it with smaller text in button

ValkoVolodya replied:
Why not to use rem on text nodes, and em on thinks, that attached to nodes? This will bring more responsibility to letter.

webknjaz replied:
Will this work for the most of email clients?

cleac replied:
@webknjaz, @ValkoVolodya I was not sure it does, so I used `px`. But the most guides used them, so I did so :smile:

anxolerd commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+        <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
+    </head>
+    <body style="marin: 8px; padding: 0px; max-width: 768px">
+        <center><img style="width: 200px" width="200px" src="/img/gdg_ukraine_logo.png"></center>
```
`src="/img/..."` This is bad. Make the url absolute (starting from https)

anxolerd replied:
By the way, you could use group-specific logo here as well.

cleac replied:
There is `<base href="https://gdg.org.ua"/>` in the headings of letter, so the absolute url is not required.

And what about group-specific logo - will think about it

anxolerd commented on src/GDGUkraine/templates/email/confirmation.html:
```diff
@@ -1,27 +1,27 @@
+    <body style="marin: 8px; padding: 0px; max-width: 768px">
+        <center><img style="width: 200px" width="200px" src="/img/gdg_ukraine_logo.png"></center>
+        <p style="margin-left: 20px; margin-bottom: 20px; font-family: Lato, Arial, sans-serif">Hey, pal!</p>
+        <p style="margin-left: 20px; font-family: Lato, Arial, sans-serif">
```
Is it really impossible to write paragraph styles only once?

webknjaz replied:
Only in case of using some tool, which would inline everything from CSS declarations into html tags. It's a must for web UI based email clients to have styles inline.

anxolerd replied:
@webknjaz: oh crap :(

anxolerd commented:
Hey, @bluebirrrrd ! We know you're a super-dupa busy person, but could you please help us with the text of confirmation letter? We really need your help and advise ;)

bluebirrrrd commented:
Hey guys, 

I'll skip the part with criticism and just modify the @cleac's version below.

`begin-part`

Thanks for registering to the event "{{ event.title }}".
Please confirm your attendance by clicking the button below. 

`
<div style="text-align:center">
  <a href="http://reg.gdg.org.ua/some/confirmation/url" style="display: block; background-color: #1c90f3; width: 100px; color: #eee; border-radius: 5px; text-decoration: none; padding: 10px; font-family: arial;">I will come</a>
</div>
`

If you have decided not to come, just ignore this letter.

`end-part`


Also, adding some time limits for confirmation would be nice. I skipped this part since we don't have any features for that at the moment.

[Advice] Please read The art of business correspondence by Sasha Karepina before writing any letters of confirmation. I believe all of you will find it very helpful for further mailing (& discussions on github as well).

webknjaz commented:
@cleac

cleac commented:
Currently have some work and univ's routine. Will do ASAP :smile:

webknjaz commented:
@cleac status?

cleac commented:
@webknjaz WIP cause too much time left from time I started this PR

anxolerd commented:
@cleac: ping
What is left to be done? Maybe @webknjaz and I can accomplish that while you are busy?

anxolerd commented:
@cleac ping

cleac commented:
Hi all, sorry for losing touch. I will finish this PR today, currently working on it :smiley:

cleac commented:
@webknjaz @anxolerd

I decided to use [mjml framework](https://mjml.io) for this. After trying several of them, it looks the best and easiest as for me.

I will make detached header with GDG logo and text which have to be in every single email, like `hi, pal` in separate mjml file, like `header.mjml` for it to be easily included into every email template file and mask all *html* files in that folder in gitignore. Here is a conceptual implementation of concept: [example](https://mjml.io/try-it-live/HJhtrwK5e). I'll make implementation in repo and push it here a bit later today

The only drawback I've found - we will have to put only _absolute_ urls in the letter.

webknjaz commented:
> The only drawback I've found - we will have to put only _absolute_ urls in the letter.

sounds ok

webknjaz commented:
@cleac I've implemented a set of `url_for()` resolvers, which produce absolute URLs, they are available in jinja2 templates

cleac commented:
> I've implemented a set of url_for() resolvers, which produce absolute URLs, they are available in jinja2 templates

@webknjaz awesome! :-)

Commit:
Implement confirmation email using mjml

How to build mjml emails:
 1. Install mjml globally from npm repository
 2. Run make mjml everytime you edit email
All the emails templates now have to be in directory
src/GDGUkraine/templates/email. If not, add directory to
EMAIL_DIRECTORY in Makefile

- Add mjml tempaltes building
- Add _gdg_logo.mjml file with gdg logo in it

cleac commented:
@webknjaz is there any doc on `url_for` you mentioned above? I'd like to use it, but don't know where should I search it :-)

webknjaz commented:
https://github.com/GDG-Ukraine/gdg.org.ua/blob/master/src/tests/test_utils.py#L16-L76
https://github.com/GDG-Ukraine/gdg.org.ua/blob/master/src/GDGUkraine/lib/utils/url.py#L201-L295

cleac commented:
It seems that thats all on task :-)
@webknjaz  @anxolerd @bluebirrrrd please, check it out

Commit:
Make building of confirmation url more beautiful

webknjaz commented on src/GDGUkraine/rest_controller.py:
```diff
@@ -378,7 +378,10 @@ def send_confirm_participants(self, id, **kwargs):
+                    'url': url_for(
+                        'Controller.Root.confirm',
+                        type_='class-based',
+                        url_params={'aes_hash': secure_id}),
```
using `url_for_class()` looks more explicit to me

cleac replied:
`url_for` looks like generalization method, so I used it. Also,  what about bug I had, it was a param name. Method was looking for name of url_param `"aes_hash"`, but I've given it an `"id"`.

Maybe the `url_for` should throw more explicit error instead of naked `TypeError`

webknjaz commented:
@cleac any tests?

webknjaz commented on Makefile:
```diff
@@ -144,3 +145,6 @@ stop-prod:
 environ-regen:
 	bin/mk-environ-file.sh
+
+mjml:
```
I guess this should be a `.PHONY` target

cleac replied:
What is the .PHONY target? I will do it, just have enough idea what is it

anxolerd replied:
https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html

webknjaz replied:
@cleac long story short: normal target in makefile is some file or directory which is expected to be created by that target; if such resource exists execution of such target does nothing. Marking the target as .phony instructs make to skip such check and run the target command list any way.

webknjaz replied:
You may also want to add `front-deps` target as a dependency and extend it to install utility you need :)

webknjaz commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+      from_file="$dirname/$filename_mjml"
+      to_file="$dirname/$filename.html"
+      mjml -r $from_file -o $to_file ${minify}
+      if [ "$?" = "0" ]; then
```
`$? -eq 0`

cleac replied:
fixed

webknjaz commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+      to_file="$dirname/$filename.html"
+      mjml -r $from_file -o $to_file ${minify}
+      if [ "$?" = "0" ]; then
+        printf '.'
```
why `printf`?

cleac replied:
Printf makes easier way not to brake line after printing statement

cleac replied:
By the way, `printf` is more portable way ([hint](http://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo)) to make such things, so I'll use it.

webknjaz replied:
I vote for `echo -n .` in order to use same style across the script.

cleac replied:
@anxolerd what you think? Do `echo -n` or `printf`?

webknjaz commented on src/GDGUkraine/templates/email/_gdg_logo.mjml:
```diff
@@ -0,0 +1,6 @@
+<mj-section>
+  <mj-column>
+      <mj-image width="200" src="/img/gdg_ukraine_logo.png"></mj-image>
```
Any chance to have templated `src` value via `url_for_static`?

cleac replied:
Haven't tried it, will look through

webknjaz commented on Makefile:
```diff
@@ -19,6 +19,7 @@ TOX=$(PENV_BIN_PATH)tox
 USER_SHELL=/bin/zsh
 ACTIVATE_ENV=if test -d ./$(PENV); then . ./$(PENV)/bin/activate; fi; if test -f ./.exports; then . ./.exports; fi
 USE_NVM=if test -d ~/.nvm; then . ~/.nvm/nvm.sh; nvm use 5.0.0 ; fi
+EMAIL_DIRECTORY=src/GDGUkraine/templates/email
```
We should make `make` learn about dotenv ^^

webknjaz commented on Makefile:
```diff
@@ -144,3 +145,6 @@ stop-prod:
 	bin/mk-environ-file.sh
+
+mjml:
+	@bin/build_mjml.sh ${EMAIL_DIRECTORY}
```
I think you shouldn't suppress output here

Commit:
Minor improvements

- Add mjml_build script to .PHONY and frontend-deps targets
- Make gdg_logo take absolute url
- Make build_mjml.sh use `echo -n` instead of `printf`
- Remove build html file from diff

cleac commented:
I'll write down some tests later today

webknjaz commented on Makefile:
```diff
@@ -142,9 +143,10 @@ stop-prod:
 .env: environ-regen
 
-.PHONY: environ-regen
+.PHONY: environ-regen mjml
```
why do you want to regen mjml along with `.env`?

cleac replied:
@webknjaz : emmm... Is someone going to call `make .PHONY`? 'R 'u mad?

webknjaz replied:
oh, and this is an incorrect syntax for `.phony`

webknjaz commented on bin/build_mjml.sh:
```diff
@@ -18,11 +18,11 @@ END
 
 if [ "$1" = 'help' -o "$1" = 'h' -o "$1" = '--help' ]; then
-  echo "$USAGE"
+  echo -n "$USAGE\n"
```
This is a nonsense. Just don't use `-n` here.

anxolerd replied:
WTF? 
Do I understand that correctly that you disable default trailing newline and print yours instead? Why would you ever need that?

cleac replied:
My bad, I just did `:%s/printf/echo -n/g` :sweat_smile: :sweat_smile: :sweat_smile: :sweat_smile:

anxolerd commented on Makefile:
```diff
@@ -13,12 +13,14 @@ OPEN_URL=xdg-open
 MIGRATOR=$(PENV_BIN_PATH)alembic -c config/alembic.ini
 BWR=bower
 NPM=npm
+MJML=mjml
```
By the way I literally have no idea why have both you and @webknjaz created these aliases.
Can anyone explain that to me, please?

webknjaz replied:
:man_shrugging:

cleac replied:
@anxolerd the same reason, why there are predefined $(CC) and $(CXX) aliases. For more easy possibility to override them, if they won't be in PATH

Commit:
Minor improvements

- Fix usage of .PHONY target
- Fix `echo -n` issue

cleac commented:
@webknjaz @anxolerd  I won't have much time this week (preparing for a new position interview), please check it out and leave comments for reasonable issues. I'll add several unit tests a bit later in another PR if you don't mind :wink:

webknjaz commented on bin/build_mjml.sh:
```diff
@@ -18,11 +18,11 @@ END
 MJML_REGEX="^[^_].+\.mjml$"
 
 if [ "$1" = 'help' -o "$1" = 'h' -o "$1" = '--help' ]; then
-  echo -n "$USAGE\n"
```
Why?

cleac replied:
@webknjaz Why do I need this `USAGE`? Just for little help if someone will want to use this script in other place without reading the code. I wasn't going to add this script to the `Makefile` while writing it, so I decided to add some help and now I have no idea why I should delete it :man_shrugging:

webknjaz replied:
no, why did you remove `-n`? (again)

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+  else
+    if [ -z $dirs ]
+    then
+      dirs=$arg
```
what if `arg` contains spaces?

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+  if [ "$arg" = '-m' -o "$arg" = '--minify' ]; then
+    minify='-m'
+  else
+    if [ -z $dirs ]
```
if `dirs` contain spaces this fails with error

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+  fi
+done
+
+if [ -z $dirs ]; then 
```
if dirs contain spaces this fails with error

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+done
+
+if [ -z $dirs ]; then 
+  echo -n 'Please, specify directory, where the files to build are!\n'
```
`echo 'Please, specify ...to build are!'`

anxolerd replied:
BTW, what abound sending this to `stderr` ?

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+fi
+
+for dirname in $dirs; do
+  if [ -d $dirname ]; then
```
Fails if `dirname` contains spaces

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+for dirname in $dirs; do
+  if [ -d $dirname ]; then
+    echo -n "Building for directory $dirname "
+    for file in $(ls $dirname | grep -E $MJML_REGEX )
```
probably `$dirname` and `$MJML_REGEX` should be quoted

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+      if [ $? -eq 0 ]; then
+        echo -n '.'
+      else
+        echo -n 'ERROR\nAn error occured, please check it out\n'
```
what about sending this to `stderr` ?

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+        exit 3
+      fi
+    done
+    echo -n ' Done\n'
```
`echo ' Done'`

anxolerd commented on bin/build_mjml.sh:
```diff
@@ -0,0 +1,69 @@
+    done
+    echo -n ' Done\n'
+  else 
+    echo -n "\nDirectory '$dirname' was not found: skipping\n"
```
`echo "\nDirectory ... skipping"`

cleac commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Finish #117](https://trello.com/c/UeT4feW2/1-finish-117)

Commit:
Fix spaces issues in script running

- Make errors be thrown to stderr

cleac committed:
Merge branch 'master' into master

cleac commented:
@webknjaz @anxolerd Guys, can you pls check out this PR? I've fixed the script
